### DISCUSSION TITLE
Run from SD card without PC

### DISCUSSION THREAD

**Discussion by drom89** [terjeio/ioSender#335] [Q&A] [answered]
Hi, so i tried to run program from SD card thru ioSender for the first time.
I just double click on the file in ioSender in SDcard tab and it started to work.
After that i tried to close the ioSender, but it didnt let me close it, so i disconnected from wifi (im connected thru wifi connection).
Everything was fine, program was running but when i tried to connect again and start ioSender, the program stopped and after that it started to move in XY plane inside the material ;-(.

Can some one elaborate how to connect again without issues ? 
Should i try it to run from WebUI ?

Thanks

**Comment by terjeio** [comment]
> Can some one elaborate how to connect again without issues ?

Not possible without code changes. At least I need to check what happened on the attempted reconnect, it should not cause irregular moves.

> Should i try it to run from WebUI ?

Yes, the WebUI has (re)connect handling implemented. But only if the job was started from a WebUI client, if a different client connects the job is terminated.

**Comment by phil-barrett** [comment]
I believe ioSender forces a reset upon connection. Perhaps reconnect would be a nice feature but...

Definitely do not close ioSender during a job if you don't have a physical way to control the machine.  If nothing else, you want a way to stop the job if something goes wrong.  You definitely need an EStop. Physical cycle start and feed hold buttons are a very good idea. You might want to add something like this:
![PXL_20230822_204633621_DxO_2048](https://github.com/terjeio/ioSender/assets/16570958/8f1aa11a-ce66-4a18-ab24-f4157aaa38fd)

By the way, you might want to check out download and run - right click on the gcode file and select download and run. This allows you to to see Program view and 3D view.

**Reply by drom89** [reply]
I have EStop, but i didn't expect that move, so i wasn't that quick.. :)

I thought that download and run is same as if you just run code from PC... isn't it ?

**Reply by phil-barrett** [reply]
I'm not sure the precise difference but it does use the file on the SD Card, not one from the PC.

**Reply by terjeio** [reply, marked as answer]
> I believe ioSender forces a reset upon connection.

Only if the controller is in a state that prevents communication, and then after user confirmation.

> I thought that download and run is same as if you just run code from PC... isn't it ?

No, the code is still run off the SD card. The download is used by the sender for rendering the 3D view. Note that the program listing will not be kept in sync since no "ok" response for each block is sent to the sender. I could have added syncing if the gcode has line numbers, but this is no longer common (AFAIKT).

**Comment by mrcgrndt94** [comment]
To be sure - running from SD card and disconnecting is supposed to be working? I have a Flexihal board with a uFlexinet connected to it. I'm connected to the Flexihal via Ethernet. I already had two situations where the gcode execution just stopped in the middle of the file without an error. I thought this was related to the network card of the PC (some energy saving stuff or so...) and wanted to just stream it via the SD card. But the same issue happened today while streaming from the SD card. Like Drom I noticed that just closing ioSender doesn't seem to work. So maybe there is still a mandatory connection to the PC even while streaming from the SD?

**Reply by terjeio** [reply]
> So maybe there is still a mandatory connection to the PC even while streaming from the SD?

I does not seem so - it would depend on the sender. E.g. ioSender does not allow you to disconnect when a program is running but it keeps running if the sender is killed. The WebUI leaves it running and allows disconnect/reconnect, even from a different machine, but a connection from a different client will stop the job.

**Comment by Bartekn86** [comment]
Hey,
So maybe it's time to enable the device to work without the sender turned on? My workspace is 1m x 1m, so the computer doesn't necessarily need to be turned on so long time..

**Reply by terjeio** [reply]
[New edge release](https://github.com/terjeio/ioSender?tab=readme-ov-file#edge-pre-releases) is up that allows closing the sender.

**Reply by Bartekn86** [reply]
thankk youuuuuuuuuuuuu veryyyyyyy muchhhhhhhhh ;-))))))))))))))

**Reply by Bartekn86** [reply]
Hi, is ther any posibility to format sdcard by iosender (tenssy 4.1) to litlefs ?

**Reply by terjeio** [reply]
> Hi, is ther any posibility to format sdcard by iosender (tenssy 4.1) to litlefs ?

[It seems it is possible](https://duckduckgo.com/?t=ffab&q=littlefs+sd+card&ia=web), but will take some work to add. Any good reason why?

BTW the iMXRT1062 driver supports littlefs in flash (8MB for 4.1) if that is enough - currently available if you enable the WebUI in the Web Builder or without the WebUI if you compile yourself.

**Comment by Bartekn86** [comment]
Hey, thanks again for the reply... it doesn't actually have to be litlefs, it's just about formatting the card itself, and the file system doesn't matter, as long as iOSender sees it.

I assumed this option would be useful.

If it requires any more work, then it probably doesn't make sense.

**Reply by terjeio** [reply]
I have added the command `$FF=yes` for formatting the card from the MDI. I have not checked if it is available for all drivers yet, likely not since it relies on a configuration of the 3rd party FatFS driver I use. If not available the command will return an error,

**Reply by Bartekn86** [reply]
thank you, waiting to update of webbuilder to test it, for now the tennsy get an error

**Reply by Bartekn86** [reply]
$FF=yes
error:2 - Missing the expected G-code word value or numeric value format is not valid.
$fm
$FF=yes
error:2 - Missing the expected G-code word value or numeric value format is not valid.
latest pre sender
teensy4.1

If you could tell me what i doing wrong..  thank you for that and for great job 
I'm very impressed with the amount of work you did, and I really liked the joke about eating and sleeping ;p :-)

**Reply by terjeio** [reply]
There were two issues with the [third party FatFs driver](https://github.com/grblHAL/uSDFS/tree/master) - should be ok now. At least it works with the 2 Gb card I use for testing.
